### ISSUE TITLE
09-k-means-housing.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/sarayruiz---Machine-Learning#8]
Hola @sarayruiz , por aquí las issues del ejercicio de clustering!

**Saray**, como en ejercicios anteriores un análisis muy completo y buena capacidad de autocrítica. El loop probando diferentes valores de k con la gráfica comparativa es genial. Las conclusiones finales están muy bien estructuradas. Sin embargo, tienes dos errores importantes: el `make_blobs` que genera datos aleatorios (no representa tu dataset) y el **data leakage** al escalar antes del split. Por lo demás, enhorabuena.

## LO QUE HAS HECHO BIEN

- **KElbowVisualizer**: Excelente uso de Yellowbrick para el método del codo
- **K-Means con k=6**: Valor correcto
- **Visualización geográfica**: Scatter plot de Lon vs Lat con clusters
- **Histograma de MedInc por cluster**: Muy buena visualización para entender la segmentación
- **Observaciones interpretativas**: Buena explicación de cómo agrupa por ingresos y ubicación
- **KNeighborsClassifier**: Opción interesante y bien justificada (relaciones espaciales)
- **Accuracy train y test**: Verificas ambos para detectar overfitting
- **Matriz de confusión**: Visualizada con heatmap
- **Classification report**: Completo, en formato DataFrame
- **Loop de k valores**: Excelente análisis para encontrar el mejor k
- **Gráfica Accuracy vs k**: Muy buena visualización
- **RandomizedSearchCV**: Optimización completa con múltiples hiperparámetros
- **Autocrítica sobre overfitting**: Muy bien identificado, demuestras comprensión
- **Guardas K-Means Y el clasificador**: Correcto
- **Conclusiones finales EXCELENTES**: Muy completas y bien estructuradas

---

### 1. ERROR: make_blobs no tiene sentido aquí

<img width="1066" height="580" alt="Image" src="https://github.com/user-attachments/assets/f5ae228a-de02-4f20-b2e9-0d6d501ee806" />

`make_blobs` **genera datos sintéticos aleatorios**, no tiene nada que ver con tu dataset de California Housing. Este gráfico no representa tus datos reales :). Deberías  reemplazarlo por:

```python
# Visualización de datos REALES antes de escalar
plt.scatter(df["Longitude"], df["Latitude"])
plt.title("Distribución geográfica de viviendas")
plt.xlabel("Longitude")
plt.ylabel("Latitude")
plt.show()
```

---

### 2. ERROR: Escalas TODO el dataset ANTES del split (DATA LEAKAGE)

Veo que estás escalando para el elbow QUE ES CORRECTO PERO el problema es que es antes del split. Lo que has hecho primero:

```python
min_max_scaler = MinMaxScaler()
min_max_features = min_max_scaler.fit_transform(df)  # ← Escalas TODO
df_escalado = pd.DataFrame(min_max_features, ...)
```

Y luego haces el split sobre los datos ya escalados. El scaler "ve" los datos de test al calcular min/max.

**La forma correcta:**
```python
# 1. Split PRIMERO (datos sin escalar)
X_train_raw, X_test_raw = train_test_split(df, test_size=0.2, random_state=18)

# 2. Escalar solo train
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train_raw)

# 3. Método del codo con train escalado
elbow_vis = KElbowVisualizer(KMeans(random_state=18), k=(2, 20))
elbow_vis.fit(X_train_scaled)

# 4. K-Means con train escalado
kmeans = KMeans(n_clusters=6, random_state=18)
kmeans.fit(X_train_scaled)

# 5. Escalar test (solo transform, no fit)
X_test_scaled = scaler.transform(X_test_raw)
```

¿Por qué importa?
Cuando haces fit_transform en todo el dataset:

- El scaler calcula min/max usando también los datos de test
- Esto es "ver las respuestas del examen" antes de hacerlo
- En este ejercicio el impacto es mínimo, pero hay quien lo considera mala práctica
---

### 3. Sugerencia: Falta Silhouette Score

Lo importas pero no lo usas:

Añade:
```python
silhouette = silhouette_score(df_escalado.drop(columns=['cluster']), 
                               df_escalado['cluster'])
print(f"Silhouette Score: {silhouette:.4f}")
```

## Silhouette Score - Explicación

El **Silhouette Score** (coeficiente de silueta) es una métrica para evaluar la **calidad del clustering**. Responde a la pregunta: **¿Qué tan bien separados están los clusters?**

### ¿Qué mide?

Para cada punto, calcula dos cosas:

1. **a** = Distancia media a los demás puntos de **su mismo cluster** (cohesión interna)
2. **b** = Distancia media a los puntos del **cluster más cercano** (separación externa)

### ¿Cómo interpretar el resultado?

| Valor | Significado |
|-------|-------------|
| **+1** | Perfecto: clusters muy densos y bien separados |
| **+0.5 a +1** | Buena estructura de clusters |
| **+0.25 a +0.5** | Estructura débil, posible solapamiento |
| **~0** | Clusters solapados, puntos en la frontera |
| **Negativo** | Puntos asignados al cluster incorrecto |

### ¿Por qué usarlo?

1. **Complementa el método del codo**: El codo te dice cuántos clusters, el silhouette te dice si esos clusters tienen sentido

2. **Detecta problemas**: Si el silhouette es bajo, quizás:
   - Elegiste mal el número de clusters
   - Los datos no tienen estructura de clusters natural
   - Algunas variables no aportan a la separación

**En resumen**: El método del codo te ayuda a elegir **cuántos clusters**, el Silhouette Score te dice **qué tan buenos son esos clusters**. Son complementarios.

---

### 3. MUY LEVE: Falta guardar el Scaler

```python
with open('../models/09-housing-modelknn.pkl', 'wb') as f:
    pickle.dump(model, f)

with open('../models/09-housing-modelkmeans.pkl', 'wb') as f:
    pickle.dump(kmeans, f)

# FALTA:
with open('../models/09-housing-scaler.pkl', 'wb') as f:
    pickle.dump(min_max_scaler, f)
```

Sin el scaler no podrías transformar nuevos datos correctamente.

---
